### PR TITLE
ENG-16311-add-argument-validation-in-TraceAgnet

### PR DESCRIPTION
**Pull request by victorxgao**

### CODE REVIEW

File: src/frontend/org/voltdb/TraceAgent.java [hunk 1]
```diff
@@ -68,28 +69,57 @@ private String parseParamsForSystemInformation(ParameterSet params, JSONObject o
...
         }
-        // Would be nice to have subselector validation here, maybe.  Maybe later.
-        obj.put("subselector", subselector);
-        if (params.toArray().length >= 2) {
-            obj.put("categories", params.toArray()[1]);
+        
+        // check number of arguments and then check the validity of arguments
+        if (subselector.equalsIgnoreCase("status") || subselector.equalsIgnoreCase("dump")) {
+            if (params.toArray().length != 1) {
+                return "Incorrect number of arguments to @Trace (expected: 1,  received: " +
```
Comment (rdykiel) [inline src/frontend/org/voltdb/TraceAgent.java]: Nit: Do you want to have the subselector value in the error message

File: src/frontend/org/voltdb/TraceAgent.java [hunk 2]
```diff
@@ -68,28 +69,57 @@ private String parseParamsForSystemInformation(ParameterSet params, JSONObject o
...
             }
         }
-        // Would be nice to have subselector validation here, maybe.  Maybe later.
-        obj.put("subselector", subselector);
-        if (params.toArray().length >= 2) {
-            obj.put("categories", params.toArray()[1]);
+        
+        // check number of arguments and then check the validity of arguments
+        if (subselector.equalsIgnoreCase("status") || subselector.equalsIgnoreCase("dump")) {
+            if (params.toArray().length != 1) {
```
Comment (rdykiel) [inline src/frontend/org/voltdb/TraceAgent.java]: You might want to do params.toArray() only once in this method. Not a big deal so no need to change, but think about the cost of such repeated calls when you do other code later.

File: src/frontend/org/voltdb/TraceAgent.java [hunk 3]
```diff
@@ -68,28 +69,57 @@ private String parseParamsForSystemInformation(ParameterSet params, JSONObject o
...
+            for (VoltTrace.Category cat : VoltTrace.Category.values()) {
+                if (cat.toString().equalsIgnoreCase(categoryItem)) {
+                    obj.put("subselector", subselector);
+                    obj.put("categories", params.toArray()[1]);
+                    obj.put("interval", false);
+                    return null;               
+                }
+            }
+
+            return "Second argument to @Trace must be a valid STRING category, instead was " +
```
Comment (rdykiel) [inline src/frontend/org/voltdb/TraceAgent.java]: You could print the VoltTrace.Categoriy.values(), I think Java will properly print the set of values

File: src/frontend/org/voltdb/TraceAgent.java [hunk 4]
```diff
@@ -60,6 +60,7 @@ protected void collectStatsImpl(Connection c, long clientHandle, OpsSelector sel
     // Parse the provided parameter set object and fill in subselector and interval into
     // the provided JSONObject.  If there's an error, return that in the String, otherwise
     // return null.  Yes, ugly.  Bang it out, then refactor later.
+    // Uppdated by Xiang Gao on Jun. 05, 2019
```
[the remark this answers is not in the document]
Comment (victorxgao) [reply src/frontend/org/voltdb/TraceAgent.java]: Sure. I will remove this line.

File: src/frontend/org/voltdb/TraceAgent.java (hunk 1 above)
Comment (victorxgao) [reply src/frontend/org/voltdb/TraceAgent.java]: Yeah. It is clearer to add the value of subselector.

File: src/frontend/org/voltdb/TraceAgent.java (hunk 2 above)
Comment (victorxgao) [reply src/frontend/org/voltdb/TraceAgent.java]: Sure. I will keep this point in mind.

File: src/frontend/org/voltdb/TraceAgent.java (hunk 3 above)
Comment (victorxgao) [reply src/frontend/org/voltdb/TraceAgent.java]: I did the test. Java did not properly print the set of Category.

File: src/frontend/org/voltdb/TraceAgent.java [hunk 5]
```diff
@@ -59,37 +59,72 @@ protected void collectStatsImpl(Connection c, long clientHandle, OpsSelector sel
...
             }
         }
-        // Would be nice to have subselector validation here, maybe.  Maybe later.
-        obj.put("subselector", subselector);
-        if (params.toArray().length >= 2) {
-            obj.put("categories", params.toArray()[1]);
+        
+        // check number of arguments and then check the validity of arguments
+        if (subselector.equalsIgnoreCase("status") || subselector.equalsIgnoreCase("dump")) {
+            if (numOfParams != 1) {
```
Comment (zeemanhe) [inline src/frontend/org/voltdb/TraceAgent.java]: The other way around is safer for comparison, that is, "status".equalsIgnoreCase(subselector)

**Pull request merged**